com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTestHelpers.cs

```diff
@@ -309,6 +309,7 @@ public static void Destroy()
             foreach (var networkManager in NetworkManagerInstances)
             {
                 networkManager.Shutdown();
+                networkManager.ShutdownInternal();
```

Comment:
😱

Reply:
To explain why this is there... It comes down to NetworkManager.Singleton.

Shutdown doesn't do anything when it's called but set a flag for later. Singleton doesn't get unset until ShutdownInternal is called. Which means timing can happen as follows:

- Test 1 runs
- Test 1 stops and shuts down the NetworkManager
- Test 2 starts running and OnSetup creates a new prefab
- NetworkManager.Singleton is still set, so the new NetworkObject gets its NetworkObject set to the current Singleton which is _already shutdown and its GameObject destroyed_
- Setup yields
- NetworkManager.ShutdownInternal is called
- NetworkManager.OnDestroy is called and destroys all NetworkObjects associated with it... _the newly created prefab gets destroyed out of nowhere!_
- Test 2 resumes and tries to use the prefab, and can't because it has now been destroyed... Test 2 fails, but if it's run on its own, it succeeds

Adding an explicit call to ShutdownInternal at the end of the test ensures the Singleton gets unset so the objects in the next test don't get associated with it.

Reply:
The ENTIRE point of the "mostly remove the singleton" work we did was to enable more reliable testing. Why are these tests using the singleton at all?

Any/all cases depending on the singleton should be considered a high priority bug IMO. Calling an internal API to work around this is just throwing more fuel on the fire.

Reply:
The tests aren't using the singleton. The SDK code is using the singleton in a way that creates non-deterministic timing artifacts that interfere with test stability.

Reply:
Kitty is correct.  The NetcodeIntegrationTest doesn't use NetworkManager.Singleton , but there are still a few tests that were not upgraded to use NetcodeIntegrationTest and/or still should be refactored to not use NetworkManager.Singleton:
_NetworkTimeSystemTests
NetworkTransformOwnershipTests
NetworkObjectSceneSerializationTests
BufferDataValidationComponent_

Regarding the SDK, there are still a few places that use it (like Kitty was mentioning)...some places are just really legacy systems (i.e. NetworkLog) and other places just need to be updated (i.e. SceneEventData line 631...not sure why that didn't get updated to use m_NetworkManager) but only a handful.